src/Elliptic/Initialization/BoundaryConditions.hpp

```diff
@@ -0,0 +1,171 @@
+
+  using sources_tag =
+      db::add_tag_prefix<Tags::Source, typename system::fields_tag>;
+
+  using simple_tags = db::AddSimpleTags<>;
+  using compute_tags = db::AddComputeTags<>;
+
+  template <typename NormalDotNumericalFluxComputer,
+            typename... NumericalFluxTags, typename... BoundaryDataTags>
+  static void compute_dirichlet_boundary_normal_dot_numerical_flux(
+      const gsl::not_null<Variables<tmpl::list<NumericalFluxTags...>>*>
+          numerical_fluxes,
+      const NormalDotNumericalFluxComputer& normal_dot_numerical_flux_computer,
+      const Variables<tmpl::list<BoundaryDataTags...>>& boundary_data,
+      const tnsr::i<DataVector, system::volume_dim, Frame::Inertial>&
+          normalized_face_normal) noexcept {
+    normal_dot_numerical_flux_computer.compute_dirichlet_boundary(
+        make_not_null(&get<NumericalFluxTags>(*numerical_fluxes))...,
+        get<BoundaryDataTags>(boundary_data)..., normalized_face_normal);
+  };
+
+  template <typename TagsList>
+  static auto initialize(
+      db::DataBox<TagsList>&& box,
+      const Parallel::ConstGlobalCache<Metavariables>& cache) noexcept {
+    static constexpr const size_t volume_dim = system::volume_dim;
+
+    const auto& analytic_solution =
+        Parallel::get<typename Metavariables::analytic_solution_tag>(cache);
+    const auto& normal_dot_numerical_flux_computer =
+        Parallel::get<typename Metavariables::normal_dot_numerical_flux>(cache);
+
+    db::mutate<sources_tag>(
+        make_not_null(&box),
+        [
+          &analytic_solution, &normal_dot_numerical_flux_computer
+        ](const gsl::not_null<db::item_type<sources_tag>*> sources,
+          const Mesh<volume_dim>& mesh,
+          const db::item_type<Tags::BoundaryDirectionsInterior<volume_dim>>&
+              boundary_directions,
+          const db::item_type<Tags::Interface<
+              Tags::BoundaryDirectionsInterior<volume_dim>,
+              Tags::Coordinates<volume_dim, Frame::Inertial>>>& boundary_coords,
+          const db::item_type<Tags::Interface<
+              Tags::BoundaryDirectionsInterior<volume_dim>,
+              Tags::Normalized<Tags::UnnormalizedFaceNormal<volume_dim>>>>&
+              normalized_face_normals,
+          const db::item_type<Tags::Interface<
+              Tags::BoundaryDirectionsInterior<volume_dim>,
+              Tags::Magnitude<Tags::UnnormalizedFaceNormal<volume_dim>>>>&
+              magnitude_of_face_normals) noexcept {
+          // Impose Dirichlet boundary conditions as contributions to the source
+          for (const auto& direction : boundary_directions) {
+            const size_t dimension = direction.dimension();
+            // Is it worth it getting this from the `mortar_data_tag`?
+            const auto mortar_mesh = mesh.slice_away(dimension);
+            // Compute Dirichlet data on mortar
+            Variables<typename system::impose_boundary_conditions_on_fields>
```

Comment:
The Variables would need to be initialised with a different number of points for each direction, so not sure if re-using it would have a performance benefit. Refactoring to modify `sources` in-place seems difficult. However, I added a few `std::move`s that should make things better: `dirichlet_boundary_data` and `boundary_normal_dot_numerical_fluxes` are moved as rvalues now and `dg::lift_flux` actually does the same internally.

Reply:
Ah well, the `std::move` should help. :+1: